common: IRCv3++, more IRCv3, allow disabling IRCv3 caps

digitalcircuit opened the pull request:
## In short

* Bundle in [the PostgreSQL text char limit changes PR](https://github.com/quassel/quassel/pull/478 )
  * That PR has already been tested
  * Simplifies merging
* Bundle in the remaining changes [from justJanne's IRCv3 PR](https://github.com/quassel/quassel/pull/479 )
  * Add [`setname`](https://ircv3.net/specs/extensions/setname ) support, now graduated from draft
  * Add [`echo-message`](https://ircv3.net/specs/extensions/echo-message-3.2.html ) support, **disabled by default**, lays groundwork for [`labeled-response`](https://ircv3.net/specs/extensions/labeled-response )
  * Fix up minor issues
* Allow skipping negotiation of IRCv3 capabilities via `Features` tab, `SkipIrcCaps` protocol flag
  * If skipped, Quassel won't request the capability, but it will handle the user `/cap req`'ng it
  * Provides escape hatch for misbehaving networks or if a bug is found in Quassel
  * Add checkbox for `server-time` in Features, for networks with wrong time
  * Add `Advanced...` button to directly set/clear list of skipped capabilities
  * Add `Details...` button to try to explain IRCv3 caps, also show currently enabled caps
* Add additional IRCv3 capabilities
  * [`account-tag`](https://ircv3.net/specs/extensions/account-tag-3.2 ) to get account information with every message
  * [`invite-notify`](https://ircv3.net/specs/extensions/invite-notify-3.2 ) to unify `/invite` messages from the server
  * Request [`message-tags`](https://ircv3.net/specs/extensions/message-tags ) - already supported, just not negotiated
* Switch infobar `dialog-*` icons to `emblem-*`
  * More distinguishable at the small sizes used
  * Falls back to `dialog-*` when unavailable (e.g. Oxygen)
* Additional cleanups
  * Use existing `IrcTags::SERVER_TIME` for `server-time` handling
  * Only use timestamps in `time` tag when `server-time` is enabled
  * Mention serialization verification in SQL database/protocol `README.md`

Criteria | Rank | Reason
---------|---------|-------------
Impact | ★★☆ *2/3* | User control over IRCv3 features, better support for modern networks
Risk | ★★☆ *2/3* | Schema change limits downgrade, possible wrong IRC implementation
Intrusiveness | ★★★ *3/3* | UI, string, database, protocol changes, will interfere with adjacent PRs

## Rationale

As Quassel adds support for more modern IRCv3 features, the risk increases of running into networks that misbehave, or that the IRCv3 capability wasn't implemented correctly in Quassel.  To handle these cases, Quassel should have a way to skip/disable negotiation of specific IRCv3 capabilities.

Of particular note, [`server-time`](https://ircv3.net/specs/extensions/server-time-3.2 ) replaces Quassel's timestamps for messages with whatever the network specifies.  If the IRC network isn't configured with the correct time (e.g. `ntp`), Quassel will appear broken, showing the wrong timestamps.  Given this risk, Quassel should make disabling `server-time` easier, and any future major IRCv3 changes (e.g. `echo-message` + `labeled-response`).

Pre-0.14 Quassel clients won't be able to modify the list of skipped capabilities, but post-0.14, the `Advanced...` button provides a way to reset to defaults.

*Check the individual commit messages for more details.*

## Examples
### `Features` tab in `Networks`
#### `Features` tab with a modern core
*In Settings → IRC → Networks, showing the `Features` tab on a modern Quassel core supporting `SkipIrcCaps`*
![Settings - IRC - Networks settings page showing the "Features" tab](https://zorro.casa/sync/Hosting/Utilities/Quassel/Development/pr/ft-ircv3-specced-out/Network%20settings%20-%20Features%20-%20modern%20core.png#v2 )
> *:information_source: These features require support from the network - `Details...` button*
>
> \[checkbox\] Use network timestamps for messages
>
> `Advanced...` button
>
> *Changes apply after reconnecting to the network*

> Tooltip text for *Use network timestamps for messages*
> Let the IRC network control the timestamps for messages, such as when viewing messages from the past.
>
> This is useful when connecting to bouncers, IRC bridges, or networks that send messages you've missed when reconnecting.
>
> *Toggles the IRCv3 'server-time' capability.*

#### `Details...` with supported core on `Features` tab, `freenode`
*In Settings → IRC → Networks, clicking the `Details...` button on the `Features` tab on a modern Quassel core connected to `freenode`*
![Dialog "Configuring network features" shown](https://zorro.casa/sync/Hosting/Utilities/Quassel/Development/pr/ft-ircv3-specced-out/Network%20settings%20-%20Features%20-%20details%20on%20freenode.png#v1 )
> Dialog *Configuring network features*:
> Quassel makes use of newer IRC features when supported by the IRC network. If desired, you can disable unwanted or problematic features here.
>
> The [IRCv3 website](https://ircv3.net/irc/ ) provides more technical details on the IRCv3 capabilities powering these features.
>
> Currently enabled IRCv3 capabilities for this network: `account-notify`, `away-notify`, `cap-notify`, `chghost`, `extended-join`, `multi-prefix`

If the network is disconnected or no IRCv3 features are enabled, the `Currently enabled …` part of the dialog is omitted.

#### `Advanced...` edit on `Features` tab
*In Settings → IRC → Networks, clicking the `Advanced...` button on the `Features` tab*
![Dialog "Edit Network Features" shown](https://zorro.casa/sync/Hosting/Utilities/Quassel/Development/pr/ft-ircv3-specced-out/Network%20settings%20-%20Features%20-%20advanced%20edit.png#v1 )
> Dialog *Edit Network Features*:
> Enter a space-separated list of IRCv3 capabilities to ignore:
>
> \[empty single line textbox\]
>
> A list of capabilities may be found at https://ircv3.net/irc/.
>
> `Defaults` button, greyed out - `OK` button - `Cancel` button

The `Defaults` button clears the textbox, providing an explicit way to return Quassel to the default state.

#### Unsupported, old core
*In Settings → IRC → Networks, showing the `Features` tab on an older Quassel core **not** supporting `SkipIrcCaps`*
![Settings - IRC - Networks settings page showing a note inside under the "Features" tab.  Below this, a dialog for "Configuring network features unsupported" is shown](https://zorro.casa/sync/Hosting/Utilities/Quassel/Development/pr/ft-ircv3-specced-out/Network%20settings%20-%20Features%20-%20unsupported.png#v3 )
> *✖️ Your Quassel core is too old to configure features - `Details...` button*

> Dialog *Configuring network features unsupported*:
> **Your Quassel core is too old to configure network features**
> You need a Quassel core v0.14.0 or newer to control what network features Quassel will use.

### New emblem icons
#### `Use SASL Authentication` status
*In Settings → IRC → Networks, showing the `Auto Identify` tab*

**Before, `Use SASL Authentication`**
![Settings - IRC - Networks settings page showing the "Use SASL Authentication" status icons for supported, unsupported, disconnected, and unknown - info icon, warning icon, info icon, then info icon](https://zorro.casa/sync/Hosting/Utilities/Quassel/Development/pr/ft-ircv3-specced-out/Icons%20-%20Use%20SASL%20-%20before.png#v1 )

**After, `Use SASL Authentication`**
![Settings - IRC - Networks settings page showing the "Use SASL Authentication" status icons for supported, unsupported, disconnected, and unknown - checkmark icon, not available icon, question icon, then question icon](https://zorro.casa/sync/Hosting/Utilities/Quassel/Development/pr/ft-ircv3-specced-out/Icons%20-%20Use%20SASL%20-%20after.png#v1 )

#### `Local Highlights` device-only notice
*In Settings → IRC → Local Highlights, at the top*

**Before, `Local Highlights apply to this device only`**
![Settings - IRC - Local Highlights settings page showing the "Local Highlights apply to this device only" with a black-and-white dialog information icon](https://zorro.casa/sync/Hosting/Utilities/Quassel/Development/pr/ft-ircv3-specced-out/Icons%20-%20Local%20Highlights%20-%20before.png#v1 )

**After, `Local Highlights apply to this device only`**
![Settings - IRC - Local Highlights settings page showing the "Local Highlights apply to this device only" with a blue information icon](https://zorro.casa/sync/Hosting/Utilities/Quassel/Development/pr/ft-ircv3-specced-out/Icons%20-%20Local%20Highlights%20-%20after.png#v1 )

#### `Remote Highlights` unsupported notice
*In Settings → IRC → Remote Highlights with an older Quassel core (before v0.13), at the top*

**Before, `Your Quassel core is too old to support remote highlights`**
![Settings - IRC - Remote Highlights settings page showing the "Your Quassel core is too old to support remote highlights" message with a vague, yellow dialog warning icon](https://zorro.casa/sync/Hosting/Utilities/Quassel/Development/pr/ft-ircv3-specced-out/Icons%20-%20Remote%20Highlights%20unsupported%20-%20before.png#v1 )

**After, `Your Quassel core is too old to support remote highlights`**
![Settings - IRC - Remote Highlights settings page showing the "Your Quassel core is too old to support remote highlights" message with a gray not supported 'x' icon](https://zorro.casa/sync/Hosting/Utilities/Quassel/Development/pr/ft-ircv3-specced-out/Icons%20-%20Remote%20Highlights%20unsupported%20-%20after.png#v1 )

#### `Core Information` active sessions unsupported notice
*In File → Core Info… with an older Quassel core (before v0.13)*

**Before, `Your Quassel core is too old to show active sessions`**
![File → Core Info... dialog showing the "Your Quassel core is too old to show active sessions" message with a vague, yellow dialog warning icon](https://zorro.casa/sync/Hosting/Utilities/Quassel/Development/pr/ft-ircv3-specced-out/Icons%20-%20Core%20Info%20unsupported%20-%20before.png#v1 )

**After, `Your Quassel core is too old to show active sessions`**
![File → Core Info... dialog showing the "Your Quassel core is too old to show active sessions" message with a gray not supported 'x' icon](https://zorro.casa/sync/Hosting/Utilities/Quassel/Development/pr/ft-ircv3-specced-out/Icons%20-%20Core%20Info%20unsupported%20-%20after.png#v1 )

# Testing - ☑ pass - 2020-9-3

## Summary
* All IRCv3 capabilities tested work on at minimum of InspIRCd, some were tested across other servers
* Database changes for `skipcaps` successful on SQLite and PostgreSQL, Windows and Linux
* All icons properly show up on Kubuntu 20.04 and Windows with Breeze and Breeze Dark

## Notes
* As part of testing Quassel's IRCv3 support, uncovered and helped fix [a `setname` issue in InspIRCd](https://github.com/inspircd/inspircd/commit/bc444b242fabb8ae8b5ca58c2c903b75c02e9756 ) and [an `account-tag` issue in Oragono](https://github.com/oragono/oragono/issues/1254 )!

## IRCv3 capability verification - ☑ pass

### [`setname`](https://ircv3.net/specs/extensions/setname ) with `/setname`

#### Steps

1.  Connect to an IRC network supporting `setname`
    * InspIRCd v3.7.0 and higher has it
2.  Connect to the IRC network with another IRC client
3.  In Quassel, observe the `Realname` of the second client
4.  As the second client, issue a `/setname` command
    * E.g. `/setname Rick Astley`
5.  In Quassel, observe the `Realname` of the second client
    * `Realname` should be updated
6.  As Quassel, issue a `/setname` command
    * E.g. `/setname Bruce Wayne`
7.  Verify Quassel accepts the command and properly updates the realname

#### Test results - ☑ pass

* `SETNAME` properly propagates
* `/setname` works

```
2020-08-31 23:15:46 [Debug] IRC net  14 >> "SETNAME :Bruce Wayne"
2020-08-31 23:15:46 [Debug] IRC net  14 << "@time=2020-09-01T03:15:46.078Z :dcircuit_dev!quasseldev@hostmask SETNAME :Bruce Wayne"
2020-08-31 23:15:46 [Debug] IRC net  15 << "@time=2020-09-01T03:15:46.078Z :dcircuit_dev!quasseldev@hostmask SETNAME :Bruce Wayne"
```

### [`echo-message`](https://ircv3.net/specs/extensions/echo-message-3.2.html ), disabled by default

#### Steps

1.  Connect to an IRC network supporting `echo-message`
    * Modern InspIRCd has it
2.  Observe the capability negotiation log
    * Quassel should **not** automatically negotiate `echo-message`
3.  Issue a `/cap req echo-message` to manually request `echo-message`
4.  Issue a `/cap ls` to check currently enabled capabilities
5.  Send channel messages, private messages, CTCP ACTION (`/me`) messages
    * Quassel should behave as expected

#### Test results - ☑ pass

* `echo-message` is correctly **not** automatically negotiated
```
* Ready to negotiate (found: invite-notify, account-tag, echo-message, inspircd.org/standard-replies, chghost, server-time, account-notify, multi-prefix, away-notify, inspircd.org/poison, userhost-in-names, setname, sts, batch, message-tags, extended-join, cap-notify, labeled-response, sasl)
* Negotiating capabilities (requesting: account-notify, account-tag, away-notify, cap-notify, chghost, extended-join, invite-notify, message-tags, multi-prefix, server-time, setname, userhost-in-names)...
* Capability negotiation finished (enabled: account-notify, account-tag, away-notify, cap-notify, chghost, extended-join, invite-notify, message-tags, multi-prefix, server-time, setname, userhost-in-names)
```

* Manually requested `echo-message` behaves as expected
```
2020-09-01 23:41:26 [Debug] IRC net  3 >> "CAP req echo-message"
2020-09-01 23:41:26 [Debug] IRC net  3 << "@time=2020-09-02T03:41:26.889Z :irc.example.com CAP dcircuit_dev ACK :echo-message"
2020-09-01 23:41:28 [Debug] IRC net  3 >> "CAP ls"
2020-09-01 23:41:28 [Debug] IRC net  3 << "@time=2020-09-02T03:41:28.820Z :irc.example.com CAP dcircuit_dev ls :account-notify account-tag away-notify batch cap-notify chghost echo-message extended-join inspircd.org/poison inspircd.org/standard-replies invite-notify labeled-response message-tags multi-prefix sasl server-time setname sts userhost-in-names "
[…]
2020-09-01 23:42:23 [Debug] IRC net  3 >> "PRIVMSG #test Testing"
2020-09-01 23:42:23 [Debug] IRC net  3 << "@time=2020-09-02T03:42:23.193Z;msgid=964~1596253601~255440 :dcircuit_dev!quasseldev@hostmask.IP PRIVMSG #test :Testing"
[…]
2020-09-01 23:45:31 [Debug] IRC net  3 >> "NOTICE #test Testing"
2020-09-01 23:45:31 [Debug] IRC net  3 << "@time=2020-09-02T03:45:31.509Z;msgid=964~1596253601~255458 :dcircuit_dev!quasseldev@hostmask.IP NOTICE #test :Testing"
[…]
2020-09-01 23:46:29 [Debug] IRC net  3 >> "PRIVMSG #test :\x01""ACTION does the IRCv3 dance!\x01"
2020-09-01 23:46:30 [Debug] IRC net  3 << "@time=2020-09-02T03:46:30.021Z;msgid=964~1596253601~255464 :dcircuit_dev!quasseldev@hostmask.IP PRIVMSG #test :\x01""ACTION does the IRCv3 dance!\x01"
[…]
2020-09-01 23:47:51 [Debug] IRC net  3 >> "PRIVMSG digitalcircuit Test"
2020-09-01 23:47:51 [Debug] IRC net  3 << "@time=2020-09-02T03:47:51.024Z;msgid=964~1596253601~255470 :dcircuit_dev!quasseldev@hostmask.IP PRIVMSG digitalcircuit :Test"
```

```
> Test
[11:41:16 pm] <dcircuit_dev> Test
> /cap req echo-message
> /cap ls
[11:41:28 pm] * No additional capabilities are supported (found: invite-notify, account-tag, echo-message, inspircd.org/standard-replies, chghost, server-time, account-notify, multi-prefix, away-notify, inspircd.org/poison, userhost-in-names, setname, sts, batch, message-tags, extended-join, cap-notify, labeled-response, sasl; currently enabled: account-notify, account-tag, away-notify, cap-notify, chghost, extended-join, invite-notify, message-tags, multi-prefix, server-time, setname, userhost-in-names, echo-message)
> Testing
[11:42:23 pm] <dcircuit_dev> Testing
> /notice #test Testing
[11:45:31 pm] [dcircuit_dev] Testing
> /me does the IRCv3 dance!
[11:46:30 pm] -*- dcircuit_dev does the IRCv3 dance!
> /query digitalcircuit
[11:47:48 pm] * Starting query with digitalcircuit
> Test
[11:47:51 pm] <dcircuit_dev> Test
```

### [`account-tag`](https://ircv3.net/specs/extensions/account-tag-3.2 )

#### Steps

1.  Connect to an IRC network supporting `account-tag`
    * Modern InspIRCd has it
2.  Connect to the IRC network with another IRC client, but **do not join a shared channel**
3.  As the second IRC client, authenticate with services then send a private message to Quassel
5.  In Quassel, verify the `Account` is set properly
6.  As the second IRC client, logout from services then send a private message to Quassel
7.  In Quassel, verify the `Account` is cleared as expected

#### Test results - ☑ pass

* Quassel client correctly updates `Account` status
* `account-tag` is parsed correctly

```
2020-09-02 21:28:41 [Debug] IRC net  3 << "@time=2020-09-03T01:28:41.411Z;msgid=964~1596253601~257653;account=digitalcircuit :dcircuit_ircv3_test!kiwiirc@hostmask.IP PRIVMSG dcircuit_dev :Hello world!"
2020-09-02 21:29:06 [Debug] IRC net  3 << "@time=2020-09-03T01:29:06.155Z;msgid=964~1596253601~257660 :dcircuit_ircv3_test!kiwiirc@hostmask.IP PRIVMSG dcircuit_dev :And now I'm logged out\xE2\x80\xA6"
```

### [`invite-notify`](https://ircv3.net/specs/extensions/invite-notify-3.2 )

#### Steps

1.  Connect to an IRC network supporting `invite-notify`
    * Modern InspIRCd has it
2.  Connect to the IRC network with another IRC client
3.  As the second client, create/join a channel with permissions to send invites
4.  As the second client, `/invite [nickname of Quassel client]`
5.  In Quassel, observe the outcome of the invite notification

#### Test results - ☑ pass

IRC network status buffer:
```
[9:36:06 pm] -> dcircuit_ircv3_test invited you to channel #testier
```

Debug log:
```
2020-09-02 21:36:06 [Debug] IRC net  3 << "@time=2020-09-03T01:36:06.918Z :dcircuit_ircv3_test!kiwiirc@hostmask.IP INVITE dcircuit_dev :#testier"
```

### [`message-tags`](https://ircv3.net/specs/extensions/message-tags )

#### Steps

1.  Launch Quassel core with `--debug-irc` and `--debug-irc-parsed`, and a log file
2.  Connect Quassel client to an IRC network supporting `message-tags`
    * Modern InspIRCd has it
3.  Connect to the IRC network with another IRC client
4.  As the second client, send a long `TAGMSG` to Quassel
    * E.g...
```
/quote @+ancient-knowledge=According\sto\sall\sknown\slaws\nof\saviation,\n\n\s\s\nthere\sis\sno\sway\sa\sbee\nshould\sbe\sable\sto\sfly.\n\n\s\s\nIts\swings\sare\stoo\ssmall\sto\sget\nits\sfat\slittle\sbody\soff\sthe\sground.\n\n\s\s\nThe\sbee,\sof\scourse,\sflies\sanyway\n\n\s\s\nbecause\sbees\sdon't\scare\nwhat\shumans\sthink\sis\simpossible.\n\n\s\s\nYellow,\sblack.\sYellow,\sblack.\nYellow,\sblack.\sYellow,\sblack.\n\n\s\s\nOoh,\sblack\sand\syellow!\nLet's\sshake\sit\sup\sa\slittle.\n\n\s\s\nBarry!\sBreakfast\sis\sready!\n\n\s\s\nOoming!\n\n\s\s\nHang\son\sa\ssecond.\n\n\s\s\nHello?\n\n\s\s\n-\sBarry?\n-\sAdam?\n\n\s\s\n-\sOan\syou\sbelieve\sthis\sis\shappening?\n-\sI\scan't.\sI'll\spick\syou\sup.\n\n\s\s\nLooking\ssharp.\n\n\s\s\nUse\sthe\sstairs.\sYour\sfather\npaid\sgood\smoney\sfor\sthose.\n\n\s\s\nSorry.\sI'm\sexcited.\n\n\s\s\nHere's\sthe\sgraduate.\nWe're\svery\sproud\sof\syou,\sson.\n\n\s\s\nA\sperfect\sreport\scard,\sall\sB's.\n\n\s\s\nVery\sproud.\n\n\s\s\nMa!\sI\sgot\sa\sthing\sgoing\shere.\n\n\s\s\n-\sYou\sgot\slint\son\syour\sfuzz.\n-\sOw!\sThat's\sme!\n\n\s\s\n-\sWave\sto\sus!\sWe'll\sbe\sin\srow\s118,000.\n-\sBye!\n\n\s\s\nBarry,\sI\stold\syou,\nstop\sflying\sin\sthe\shouse!\n\n\s\s\n-\sHey,\sAdam.\n-\sHey,\sBarry.\n\n\s\s\n-\sIs\sthat\sfuzz\sgel?\n-\sA\slittle.\sSpecial\sday,\sgraduation.\n\n\s\s\nNever\sthought\sI'd\smake\sit.\n\n\s\s\nThree\sdays\sgrade\sschool,\nthree\sdays\shigh\sschool.\n\n\s\s\nThose\swere\sawkward.\n\n\s\s\nThree\sdays\scollege.\sI'm\sglad\sI\stook\na\sday\sand\shitchhiked\saround\sthe\shive.\n\n\s\s\nYou\sdid\scome\sback\sdifferent. TAGMSG :dcircuit_dev
```
5.  Open the Quassel core log with e.g. `tail --follow`
6.  Observe the debug log and Quassel client for the outcome of `TAGMSG` and any unexpected issues
7.  Send a `TAGMSG` manually to another client
    * Able to reuse the above command
8.  Observe the results, ensure the `TAGMSG` is not cut off

#### Test results - ☑ pass

Quassel handles a long `TAGMSG`:
```
2020-09-02 21:55:57 [Debug] IRC net  3 << "@time=2020-09-03T01:55:57.642Z;msgid=964~1596253601~257684;+ancient-knowledge=According\\sto\\sall\\sknown\\slaws\\nof\\saviation,\\n\\n\\s\\s\\nthere\\sis\\sno\\sway\\sa\\sbee\\nshould\\sbe\\sable\\sto\\sfly.\\n\\n\\s\\s\\nIts\\swings\\sare\\stoo\\ssmall\\sto\\sget\\nits\\sfat\\slittle\\sbody\\soff\\sthe\\sground.\\n\\n\\s\\s\\nThe\\sbee,\\sof\\scourse,\\sflies\\sanyway\\n\\n\\s\\s\\nbecause\\sbees\\sdon't\\scare\\nwhat\\shumans\\sthink\\sis\\simpossible.\\n\\n\\s\\s\\nYellow,\\sblack.\\sYellow,\\sblack.\\nYellow,\\sblack.\\sYellow,\\sblack.\\n\\n\\s\\s\\nOoh,\\sblack\\sand\\syellow!\\nLet's\\sshake\\sit\\sup\\sa\\slittle.\\n\\n\\s\\s\\nBarry!\\sBreakfast\\sis\\sready!\\n\\n\\s\\s\\nOoming!\\n\\n\\s\\s\\nHang\\son\\sa\\ssecond.\\n\\n\\s\\s\\nHello?\\n\\n\\s\\s\\n-\\sBarry?\\n-\\sAdam?\\n\\n\\s\\s\\n-\\sOan\\syou\\sbelieve\\sthis\\sis\\shappening?\\n-\\sI\\scan't.\\sI'll\\spick\\syou\\sup.\\n\\n\\s\\s\\nLooking\\ssharp.\\n\\n\\s\\s\\nUse\\sthe\\sstairs.\\sYour\\sfather\\npaid\\sgood\\smoney\\sfor\\sthose.\\n\\n\\s\\s\\nSorry.\\sI'm\\sexcited.\\n\\n\\s\\s\\nHere's\\sthe\\sgraduate.\\nWe're\\svery\\sproud\\sof\\syou,\\sson.\\n\\n\\s\\s\\nA\\sperfect\\sreport\\scard,\\sall\\sB's.\\n\\n\\s\\s\\nVery\\sproud.\\n\\n\\s\\s\\nMa!\\sI\\sgot\\sa\\sthing\\sgoing\\shere.\\n\\n\\s\\s\\n-\\sYou\\sgot\\slint\\son\\syour\\sfuzz.\\n-\\sOw!\\sThat's\\sme!\\n\\n\\s\\s\\n-\\sWave\\sto\\sus!\\sWe'll\\sbe\\sin\\srow\\s118,000.\\n-\\sBye!\\n\\n\\s\\s\\nBarry,\\sI\\stold\\syou,\\nstop\\sflying\\sin\\sthe\\shouse!\\n\\n\\s\\s\\n-\\sHey,\\sAdam.\\n-\\sHey,\\sBarry.\\n\\n\\s\\s\\n-\\sIs\\sthat\\sfuzz\\sgel?\\n-\\sA\\slittle.\\sSpecial\\sday,\\sgraduation.\\n\\n\\s\\s\\nNever\\sthought\\sI'd\\smake\\sit.\\n\\n\\s\\s\\nThree\\sdays\\sgrade\\sschool,\\nthree\\sdays\\shigh\\sschool.\\n\\n\\s\\s\\nThose\\swere\\sawkward.\\n\\n\\s\\s\\nThree\\sdays\\scollege.\\sI'm\\sglad\\sI\\stook\\na\\sday\\sand\\shitchhiked\\saround\\sthe\\shive.\\n\\n\\s\\s\\nYou\\sdid\\scome\\sback\\sdifferent. :dcircuit_ircv3_test!kiwiirc@hostmask.IP TAGMSG :dcircuit_dev"
2020-09-02 21:55:57 [Debug] IRC net  3 << QHash(("(clientTag = 0, vendor = ,key = msgid", "964~1596253601~257684")("(clientTag = 1, vendor = ,key = ancient-knowledge", "According to all known laws\nof aviation,\n\n  \nthere is no way a bee\nshould be able to fly.\n\n  \nIts wings are too small to get\nits fat little body off the ground.\n\n  \nThe bee, of course, flies anyway\n\n  \nbecause bees don't care\nwhat humans think is impossible.\n\n  \nYellow, black. Yellow, black.\nYellow, black. Yellow, black.\n\n  \nOoh, black and yellow!\nLet's shake it up a little.\n\n  \nBarry! Breakfast is ready!\n\n  \nOoming!\n\n  \nHang on a second.\n\n  \nHello?\n\n  \n- Barry?\n- Adam?\n\n  \n- Oan you believe this is happening?\n- I can't. I'll pick you up.\n\n  \nLooking sharp.\n\n  \nUse the stairs. Your father\npaid good money for those.\n\n  \nSorry. I'm excited.\n\n  \nHere's the graduate.\nWe're very proud of you, son.\n\n  \nA perfect report card, all B's.\n\n  \nVery proud.\n\n  \nMa! I got a thing going here.\n\n  \n- You got lint on your fuzz.\n- Ow! That's me!\n\n  \n- Wave to us! We'll be in row 118,000.\n- Bye!\n\n  \nBarry, I told you,\nstop flying in the house!\n\n  \n- Hey, Adam.\n- Hey, Barry.\n\n  \n- Is that fuzz gel?\n- A little. Special day, graduation.\n\n  \nNever thought I'd make it.\n\n  \nThree days grade school,\nthree days high school.\n\n  \nThose were awkward.\n\n  \nThree days college. I'm glad I took\na day and hitchhiked around the hive.\n\n  \nYou did come back different.")("(clientTag = 0, vendor = ,key = time", "2020-09-03T01:55:57.642Z")) "dcircuit_ircv3_test!kiwiirc@hostmask.IP" "TAGMSG" ("dcircuit_dev")
```

*Note: `QString` encodes newlines as `\n` in `QDebug` output - Quassel is handling `\\n` escapes in `IrcDecoder::parseTagValue(const QString& value)`*

KiwiIRC receives a long `TAGMSG` back from Quassel:
```
[S] @time=2020-09-03T02:04:08.976Z;msgid=964~1596253601~257688;+ancient-knowledge=According\sto\sall\sknown\slaws\nof\saviation,\n\n\s\s\nthere\sis\sno\sway\sa\sbee\nshould\sbe\sable\sto\sfly.\n\n\s\s\nIts\swings\sare\stoo\ssmall\sto\sget\nits\sfat\slittle\sbody\soff\sthe\sground.\n\n\s\s\nThe\sbee,\sof\scourse,\sflies\sanyway\n\n\s\s\nbecause\sbees\sdon't\scare\nwhat\shumans\sthink\sis\simpossible.\n\n\s\s\nYellow,\sblack.\sYellow,\sblack.\nYellow,\sblack.\sYellow,\sblack.\n\n\s\s\nOoh,\sblack\sand\syellow!\nLet's\sshake\sit\sup\sa\slittle.\n\n\s\s\nBarry!\sBreakfast\sis\sready!\n\n\s\s\nOoming!\n\n\s\s\nHang\son\sa\ssecond.\n\n\s\s\nHello?\n\n\s\s\n-\sBarry?\n-\sAdam?\n\n\s\s\n-\sOan\syou\sbelieve\sthis\sis\shappening?\n-\sI\scan't.\sI'll\spick\syou\sup.\n\n\s\s\nLooking\ssharp.\n\n\s\s\nUse\sthe\sstairs.\sYour\sfather\npaid\sgood\smoney\sfor\sthose.\n\n\s\s\nSorry.\sI'm\sexcited.\n\n\s\s\nHere's\sthe\sgraduate.\nWe're\svery\sproud\sof\syou,\sson.\n\n\s\s\nA\sperfect\sreport\scard,\sall\sB's.\n\n\s\s\nVery\sproud.\n\n\s\s\nMa!\sI\sgot\sa\sthing\sgoing\shere.\n\n\s\s\n-\sYou\sgot\slint\son\syour\sfuzz.\n-\sOw!\sThat's\sme!\n\n\s\s\n-\sWave\sto\sus!\sWe'll\sbe\sin\srow\s118,000.\n-\sBye!\n\n\s\s\nBarry,\sI\stold\syou,\nstop\sflying\sin\sthe\shouse!\n\n\s\s\n-\sHey,\sAdam.\n-\sHey,\sBarry.\n\n\s\s\n-\sIs\sthat\sfuzz\sgel?\n-\sA\slittle.\sSpecial\sday,\sgraduation.\n\n\s\s\nNever\sthought\sI'd\smake\sit.\n\n\s\s\nThree\sdays\sgrade\sschool,\nthree\sdays\shigh\sschool.\n\n\s\s\nThose\swere\sawkward.\n\n\s\s\nThree\sdays\scollege.\sI'm\sglad\sI\stook\na\sday\sand\shitchhiked\saround\sthe\shive.\n\n\s\s\nYou\sdid\scome\sback\sdifferent. :dcircuit_dev!quasseldev@hostmask.IP TAGMSG :dcircuit_ircv3_test
```

*Note: This is KiwiIRC's raw output.  I'm not sure how to view the `message-tags` parsed output, but that doesn't matter anyways, this is ensuring the message gets sent without Quassel incorrectly cutting it short.*

## `SkipIrcCaps` database/protocol changes - ☑ pass

### Steps

1.  Initialize or upgrade Quassel core (*completing first run wizard if needed*)
2.  Verify networks intact
3.  Connect to a network supporting `server-time`, verify message sends
4.  Verify all supported IRCv3 capabilities are negotiated
    * `sasl` won't be negotiated unless `Use SASL Authentication` is enabled
5.  For new client + old core, ensure core upgrade warning is shown
    * With new client + old core, testing is complete
6.  For old client + new core, ensure settings UI works, any disabled caps are shown in Status Buffer
    * With old client + new core, testing is complete
7.  For new client + new core, uncheck `Use network timestamps for messages`, apply and reconnect
8.  Restart the core
9.  Verify networks intact
10. For new client + new core, and migration, verify
    * Connection log shows a message about `server-time` being skipped
    * `Use network timestamps for messages` remains unchecked

### Test results - ☑ pass
### Linux - ☑ pass
#### Fresh configuration - ☑ pass
*Quassel core from this PR set up without any initial data*

Test conditions  | Result  | Remarks
-----------------|---------|-------------
SQLite, new core, new client | ☑ *pass* |
SQLite, new core, old client | ☑ *pass* |
SQLite, old core, new client | ☑ *pass* |
SQLite, new monolithic | ☑ *pass* |
Postgres, new core, new client | ☑ *pass* |
Postgres, new core, old client | ☑ *pass* |
Postgres, old core, new client | ☑ *pass* |

#### Migration - ☑ pass
*Quassel core from this PR set up without any initial data, then migrated*

Test conditions  | Result  | Remarks
-----------------|---------|-------------
SQLite → Postgres, new core, new client | ☑ *pass* | *migrated from new core + new client, with + without `skipcaps` set*

#### Upgrading existing - ☑ pass
*Quassel core 0.12.5 set up, then upgraded to this PR*

Test conditions  | Result  | Remarks
-----------------|---------|-------------
SQLite, old → new core | ☑ *pass* | *upgraded from old core + new client*
SQLite, old monolithic → new monolithic | ☑ *pass* |
Postgres, old → new core | ☑ *pass* | *upgraded from old core + new client*

### Windows - ☑ pass
#### Fresh configuration - ☑ pass
*Quassel core from this PR set up without any initial data*

Test conditions  | Result  | Remarks
-----------------|---------|-------------
SQLite, new core, new client | ☑ *pass* |
SQLite, new core, old client | ☑ *pass* |
SQLite, old core, new client | ☑ *pass* |
SQLite, new monolithic | ☑ *pass* | *skipped IRCv3 testing for time reasons*

#### Migration - ❓ N/A
*Without Postgres on Windows, migration could not be tested*

#### Upgrading existing - ☑ pass
*Quassel core 0.14-git-master set up, then upgraded to this PR*

Test conditions  | Result  | Remarks
-----------------|---------|-------------
SQLite, old → new core | ☑ *pass* |
SQLite, old monolithic → new monolithic | ☑ *pass* |

### Failed cases
None found!

## Icon changes - ☑ pass

### Steps

*Note: As part of getting the above screenshots, the icons elsewhere were verified on Kubuntu 20.04 LTS.  This details the procedures for testing the icon bundling on Windows.*

1.  Launch client
2.  Set icon theme to `Breeze` in Settings → Interface
3.  Navigate to Settings → IRC → Networks, selecting the `Auto Identify` tab
4.  Add a network **supporting SASL**
    * E.g. `chat.freenode.net`, port `6697` with `Use encrypted connection` checked
5.  Select the network, observe the SASL status
    * Should say `Could not check if supported by network` with a question mark icon (❓)
6.  Save changes, close settings
7.  Re-open settings (as per step 3)
8.  Select the network, observe the SASL status
    * Should say `Cannot check if supported when disconnected` with a question mark icon (❓)
9.  Connect to the network
10. Re-open settings (as per step 3)
11. Select the network, observe the SASL status
    * Should say `Supported by network` with a check mark icon (☑)
12. Add a network **not supporting SASL**
    * E.g. `irc.oftc.net`, port `6697` with `Use encrypted connection` checked
13. Save changes, close settings
14. Connect to the network
15. Re-open settings (as per step 3)
16. Select the network, observe the SASL status
    * Should say `Not currently supported by network` with a gray 'x' not available icon (❌)
17. Set icon theme to `Breeze Dark` in Settings → Interface
18. Repeat steps 3-16 (creating new networks or resetting existing ones as needed)

### Test results - ☑ pass

Test conditions  | Result  | Remarks
-----------------|---------|-------------
Kubuntu 20.04 LTS | ☑ *pass* |
macOS | ❓ *skipped* | Unfortunately, I don't have access to macOS
Windows | ☑ *pass* |

digitalcircuit commented:
I'll be removing `echo-message` from the default capability negotiation so it can only be enabled through a manual `/CAP REQ echo-message`, and moving it into an independent commit for easier revert if need be.

I'll also clean up the `Features` UI to remove `echo-message`.  I might split out the UI additions for `server-time` into a separate commit as well just for easier reference later (versus all the code that went into adding `skipCaps` editing support).

digitalcircuit commented:
This is ready for review!

digitalcircuit commented:
Force-pushed to rebase after recent merges, fixing merge conflicts with `quassel.h`.  No functional changes.

Sput42 commented on src/core/eventstringifier.cpp:
```diff
@@ -814,23 +819,33 @@ void EventStringifier::processCtcpEvent(CtcpEvent* e)
 
-void EventStringifier::defaultHandler(const QString& ctcpCmd, CtcpEvent* e)
-{
+void EventStringifier::defaultHandler(const QString& ctcpCmd, CtcpEvent* e) {
```
Brace jumped up a line...

Sput42 commented on src/core/eventstringifier.cpp:
```diff
@@ -814,23 +819,33 @@ void EventStringifier::processCtcpEvent(CtcpEvent* e)
+    } else {
+        // Ignore echo messages for our own answers
+        if (!e->testFlag(EventManager::Self)) {
+            displayMsg(e, Message::Server,
+                       tr("Received CTCP-%1 answer from %2: %3")
+                       .arg(e->ctcpCmd(), nickFromMask(e->prefix()),e->param()));
```
Slightly weird and confusing line breaking and spacing here..

digitalcircuit replied:
That's probably since I forgot to bump Qt Creator's shown line width from 100 columns or whatever it had been to the more recent wider width.  I'll just make the `tr("…").arg(…)` into one line.

digitalcircuit commented on src/core/corenetwork.cpp:
```diff
@@ -1253,6 +1259,26 @@ void CoreNetwork::retryCapsIndividually()
+        // Revisit this if needed.
+        QStringList sortedSkipCaps = QStringList(networkInfo().skipCaps);
+        QStringList sortedCaps = QStringList(caps());
+        sortedSkipCaps.sort();
+        sortedCaps.sort();
```
Agreed!  And I think every place that Quassel itself writes to `NetworkInfo::skipCaps`, it sorts first.  However, if someone manually modifies the database or uses a third-party client that doesn't enforce sorting, the skipped caps may be unsorted which will break the following list intersection.

If that's a risk we're willing to take (and it might be unlikely), instead, I could enforce sorting when loading from the database and skip it here.

`caps()` comes from the network unsorted, and in Quassel 0.13 I had intentionally left it unsorted so Quassel could print caps in the order returned by the network (before sorting when negotiating), just in case the order of caps become important in distinguishing buggy/broken behavior.  It's not required though.

Sput42 replied:
Would it make sense then to just sort once upon receipt, and remove all the other sorts? In that case, if order ever became important, we could care about that at the root, and would retain the order throughout the rest of the code. Arguably, we'd like to log them in the order received (or used/sorted if not imported) anyway...

digitalcircuit replied:
### Sorting `skipCaps`
For `skipCaps`, I think that makes sense, and could look like this:

1.  Modify `void Network::setSkipCaps(const QStringList& skipCaps)` to sort first before applying/syncing
2.  Modify `QDataStream& operator>>(QDataStream& in, NetworkInfo& info)` to sort on deserialization
3.  Update comment in `void NetworkInfo::skipCapsFromString(const QString& flattenedSkipCaps)` to clarify reason for sorting
4.  Remove other sorting of `skipCaps`
5.  *Optional:* Add sorting to `QString NetworkInfo::skipCapsToString() const`
     * This ensures the `Advanced…` dialog under the Features tab on the Networks settings page always shows a sorted list (subjectively looks nicer, no functional impact), and ensures it's sorted when saving to the database (doesn't matter, will be sorted on load)

If I'm reading `NetworkInfo networkInfo() const;` correctly, this should prevent code from outside of the `Network` class from modifying `skipCaps` without it being sorted.

### Sorting `caps`
For `caps` which is filled from the network, I also agree, I think - that's why I'm sorting a copy here, instead of sorting the caps stored directly in the Network class.  Quassel later displays the network-offered caps in the order the network sent them (the `Ready to negotiate (found: %1)` message).

digitalcircuit replied:
I overcomplicated this as usual :upside_down_face: 

In hindsight, I can simply sort within `Network::setSkipCaps()` and adjust `CoreNetwork` to call `Network::skipCaps()` instead of reaching into `networkInfo().skipCaps` - the only other code in `CoreNetwork` that referenced `networkInfo()` like this was the `useSASL` check, which might've been an oversight - I'll fix that up in a new commit.

I'm adding sorting to `NetworkInfo::skipCapsToString()` as well just for display purposes for the `Advanced…` edit dialog, which also results in Quassel saving `skipCaps` sorted to the database, too.

digitalcircuit replied:
Test case:

`corenetwork.cpp`
```cpp
    if (!skipCaps().isEmpty() && !caps().isEmpty()) {
        // Find the entries that are common to skipCaps() and caps().  This represents any
        // capabilities supported by the server that were skipped.

        // Both skipCaps() and caps() are already lowercase
        // std::set_intersection requires sorted lists, and we can't modify the original lists.
        //
        // skipCaps() should already be sorted.  caps() is intentionally not sorted elsewhere so
        // Quassel can show the capabilities in the order transmitted by the network.
        auto sortedCaps = caps();
        sortedCaps.sort();

        showMessage(NetworkInternalMessage(
            Message::Server,
            BufferInfo::StatusBuffer,
            "",
            tr("DEBUG TEST - skipCaps(): %1").arg(skipCaps().join(", "))
        ));

        // Find the intersection between skipped caps and server-supplied caps
        std::set_intersection(skipCaps().cbegin(), skipCaps().cend(),
                              sortedCaps.cbegin(), sortedCaps.cend(),
                              std::back_inserter(capsSkipped));
    }
```

With this SQLite edit to intentionally set out-of-order `skipCaps`:
```sql
update network set skipcaps='extended-join away-notify' where networkname='Local (Bitlbee)';
```

Results in this in the server buffer:
```
* Requesting capability list...
[localhost] BitlBee-IRCd initialized, please go on
* DEBUG TEST - skipCaps(): away-notify, extended-join
* Ready to negotiate (found: sasl, userhost-in-names, multi-prefix, away-notify, extended-join, server-time)
* Quassel is configured to ignore some capabilities (skipped: away-notify, extended-join)
* Negotiating capabilities (requesting: sasl, multi-prefix, server-time, userhost-in-names)...
```

Success! :heavy_check_mark:  And editing capabilities with the Network's `Advanced…` dialog works, too.

digitalcircuit commented on src/core/corenetwork.cpp:
```diff
@@ -1253,6 +1259,26 @@ void CoreNetwork::retryCapsIndividually()
+        // Both skipCaps and caps are already lowercase
+        // std::set_intersection requires sorted lists, and we can't modify the original lists.
+        // Revisit this if needed.
+        QStringList sortedSkipCaps = QStringList(networkInfo().skipCaps);
+        QStringList sortedCaps = QStringList(caps());
```
I think I had to create a copy of the `QStringList` in order to be able to modify/sort `caps()` (since it's a constant), and I may have been wanting to avoid modifying the original list for `networkInfo().skipCaps` since I hadn't checked if that'd have any protocol side-effects - the latter I can certainly toss out.

I'll test compiling this without the copy.  [See also the reply below](https://github.com/quassel/quassel/pull/545#discussion_r504246853).

digitalcircuit replied:
Strange, it works now.  I suspect I got mixed up when making other changes and never went back to recheck this.  Thank you!

While I was here, I cleaned up the type to `auto sorted[…] = […]` as well.  This will be part of the next push.

digitalcircuit commented on src/qtui/settingspages/networkssettingspage.cpp:
```diff
@@ -178,6 +179,20 @@ void NetworksSettingsPage::load()
+    else {
+        // Core does not IRCv3 capability skipping, show warning and disable configuration
+        ui.enableCapsConfigWidget->setEnabled(false);
+        ui.enableCapsStatusLabel->setText(tr("Your Quassel core is too old to configure features"));
```
I've distinguished between "features" and "IRCv3 capabilities", since some features (e.g. showing messages as the server sent it to others) will require multiple capabilities (`echo-message`, `labeled-response`, etc).  And I'm not sure if we'd want to put things in this list which technically aren't part of IRCv3.

That said, it seems fine to just call this "IRCv3 features" to make it more distinguishable in e.g. support requests.  The `Details` button sorta explains the difference anyways.

digitalcircuit replied:
Fixed as requested in the upcoming push!  I'll update the PR screenshot, too.

For consistency, I also adjusted the message box that shows from clicking `Details…`:
```diff
- tr("Your Quassel core is too old to configure network features")
+ tr("Your Quassel core is too old to configure IRCv3 network features")
```

justjanne commented on src/core/SQL/PostgreSQL/setup_000_quasseluser.sql:
```diff
@@ -1,7 +1,7 @@
 CREATE TABLE quasseluser (
        userid serial NOT NULL PRIMARY KEY,
-       username varchar(64) UNIQUE NOT NULL,
+       username varchar(8703) UNIQUE NOT NULL,
```
VARCHAR fields end up embedded in the row, while TEXT fields end up in an additional TOAST heap, causing slower lookup times due to an additional pointer chase

Sput42 replied:
@justjanne Did some research, seems like that is true for MySQL. The Postgres docs state that `VARCHAR(n)`, `VARCHAR` and `TEXT` are identical except for a size check when inserting into the former, but that there is otherwise no performance or storage difference. For sqlite it was hard to find information, but apparently they simply ignore the number and store `VARCHAR(n)` as `TEXT` anyway, although I found a single post claiming that the limit would be used for "query planning".

Given that my main worry was that we would be wasting space, which doesn't seem to be the case in any of the common db engines including MySQL, I don't care too much which way to go. I just found the 8703 char limit slightly obscure if you aren't familiar with IRCv3 specifications, and I wonder if we really should hardcode that (in other words, if the limit gives us any benefit).

digitalcircuit replied:
With [the latest revision of commits](https://github.com/quassel/quassel/pull/545#issuecomment-723666347 ), I've addressed this by switching from `VARCHAR(n)` to `TEXT`.

digitalcircuit commented on src/core/SQL/PostgreSQL/setup_040_network.sql:
```diff
@@ -28,5 +28,6 @@ CREATE TABLE network (
        messagerateburstsize INTEGER NOT NULL DEFAULT 5,     -- Maximum messages at once
        messageratedelay INTEGER NOT NULL DEFAULT 2200,      -- Delay between future messages (milliseconds)
        unlimitedmessagerate boolean NOT NULL DEFAULT FALSE, -- Disable rate limits
+       skipcaps text,                                       -- Space-separated IRCv3 caps to not auto-negotiate
```
Now I see what I did - I had copy-pasted from another instance of `text` in this file without stopping to see which was more correct.  Thanks for calling this out, will be fixed to `TEXT`!

digitalcircuit commented:
I've addressed all comments except for…

* The [PostgreSQL `VARCHAR` vs `TEXT` discussion](https://github.com/quassel/quassel/pull/545#discussion_r504255028)
  * Awaiting reply; justJanne explained there is a performance impact
* The [matter of sorting capabilities during negotiation vs. protocol-level](https://github.com/quassel/quassel/pull/545#discussion_r504246853 )
  * Awaiting reply; I'm unsure of the best approach with handling third-party clients

Thank you for your feedback - this should be closer to the finish line :slightly_smiling_face:

Sput42 commented:
See my comments to the two open questions. I'll wait a bit for another round of feedback (or your decision which way to go in the end), then we can finally merge this - rest is fine, thx for fixing!

digitalcircuit commented:
I've [addressed all comments as per the force-push view](https://github.com/quassel/quassel/compare/497432c094373113313f5a327b6c18ec1e012e2d..b5d72f816aa1b6fc095b3d73cde9306fc7406b62 ) except for…

* The [PostgreSQL `VARCHAR` vs `TEXT` discussion](https://github.com/quassel/quassel/pull/545#discussion_r504255028)
  * Awaiting reply; letting `@justJanne` respond first before making changes
  * Alternatively, I could make the change and wait for a comment…?

I've not thoroughly retested, but [after redoing `skipCaps` sorting](https://github.com/quassel/quassel/pull/545#discussion_r519249214 ) I've [done some fundamental checks](https://github.com/quassel/quassel/pull/545#discussion_r519250794).

I also added a minor cleanup in the form of this commit: [`core: In CoreNetwork, check Network::useSasl()`](https://github.com/quassel/quassel/pull/545/commits/54f47ec660850c8215e4ff62ad24a2e222e0deba).

justjanne commented:
@digitalcircuit as I was told yesterday (TIL), in more recent versions VARCHAR and TEXT are by default the same anyway. So it doesn't seem to matter, and if specific behaviour is needed, it can be achieved separately.

digitalcircuit commented:
I've [addressed all comments as per the force-push view](https://github.com/quassel/quassel/compare/b5d72f816aa1b6fc095b3d73cde9306fc7406b62..508613c041cd29358a954eca81a5f4cd36872689 ).  I amended [the PostgreSQL commit message accordingly](https://github.com/quassel/quassel/pull/545/commits/95f328ee8aee8a9e4b5e4ce5ff83eb06e8121358 ), too.

In this situation, I retested freshly setting up a Quassel core with a blank PostgreSQL database, then retested upgrading a Quassel core PostgreSQL database from 0.12.5 to this PR.  In both cases, it appeared successful, with messages correctly sending.

Thank you `Sput` for your review, `justJanne` for making sure assumptions were double-checked, and `jcolsen` for the initial PostgreSQL changes.  Hopefully, fingers crossed, everything is good to go!
